src/lib/dispatch/dispatch_cfg.c

```diff
@@ -0,0 +1,138 @@
+/* Copyright (c) 2001, Matej Pfajfar.
+ * Copyright (c) 2001-2004, Roger Dingledine.
+ * Copyright (c) 2004-2006, Roger Dingledine, Nick Mathewson.
+ * Copyright (c) 2007-2018, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+/**
+ * \file dispatch_cfg.c
+ * \brief Create and configure a dispatch_cfg_t.
+ *
+ * A dispatch_cfg_t object is used to configure a set of messages and
+ * associated information before creating a dispatch_t.
+ */
+
+#define DISPATCH_PRIVATE
+
+#include "orconfig.h"
+#include "lib/dispatch/dispatch_cfg.h"
+#include "lib/dispatch/dispatch_cfg_st.h"
+#include "lib/dispatch/dispatch.h"
+#include "lib/dispatch/dispatch_st.h"
+
+#include "lib/container/smartlist.h"
+#include "lib/malloc/malloc.h"
+
+/**
+ * Create and return a new dispatch_cfg_t.
+ **/
+dispatch_cfg_t *
+dcfg_new(void)
+{
+  dispatch_cfg_t *cfg = tor_malloc(sizeof(dispatch_cfg_t));
+  cfg->type_by_msg = smartlist_new();
+  cfg->chan_by_msg = smartlist_new();
+  cfg->fns_by_type = smartlist_new();
+  cfg->recv_by_msg = smartlist_new();
+  return cfg;
+}
+
+/** DOCDOC */
+#define ID_TO_VOID(id) ((void*)(uintptr_t)(id))
```

Comment:
Hm. I didn't think this was disallowed, so long as you never dereferenced the pointer.  I'll look in the spec.

Reply:
From the C99 spec, I see (section 6.3.2.3):

> An  integer  may  be  converted  to  any pointer  type.Except  as  previously  specified,  the result  is  implementation-defined,  might  not  be  correctly  aligned,  might  not  point  to  an entity of the referenced type, and might be a trap representation.
>
> Any pointer type may be converted to an integer type.Except as previously specified, the result  is  implementation-defined.If  the  result  cannot  be  represented  in  the  integer  type,the  behavior  is  undefined.The  result  need  not  be  in  the  range  of  values  of  any integer type.

C11 seems to agree.

Reply:
It looks like you're guaranteed (C99 §7.18.1.4) to be able to round-trip a `void *` through an `intptr_t`, but not explicitly vice versa. Pointer/integer conversions are implementation-defined (C99 §6.3.2.3), except for null pointer constants. Integer-to-pointer conversions are allowed to produce trap representations. I think it's possible for reading a trap representation of a pointer to cause undefined behavior even if you're not dereferencing it (C99 §6.2.6.1).

On most platforms we care about we'll probably be ok, but I think there's not an easy build-time test for the behavior we expect.

Reply:
Okay.  In that case, how about if leave this patch alone for now, but we open two more tickets to:
  a) add a unit test to make sure that this actually works as we expect, to the extent that we can spot-check, and 
  b) eventually, stop doing this behavior entirely.

Reply:
It looks like I already opened a slightly related ticket https://trac.torproject.org/projects/tor/ticket/23714

Also it looks like you already had to work around a compiler warning for this, so it might become riskier in the future.

Maybe teor's suggestion of using a union could work.

Reply:
Opened https://trac.torproject.org/projects/tor/ticket/29537 about the test. I agree that it's probably safe to leave it alone for this patch, and recognize that we should stop relying on this pattern in the future.